File: matlab/example_2_spin_1_singlet.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Author: J.P.G. van Dijk                                                 %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Cleanup
clear all;
close all;
clc;

% Global variables
global nsim;
global Nsim;
global system;
global microwave_control_1 microwave_control_2 detuning_control;
global dt;
global probability;

% Create the system
system = spine.systems.system_2_spin_1_singlet();
system.setChargingEnergy(2 * pi * 100e9);
system.setRabiFrequency(2 * pi * 1e6);
system.setTunnelControl(2 * pi * 10e6 * sqrt(2));
system.setLarmorFrequency(1, 2 * pi * (1e9 - 10e6));
system.setLarmorFrequency(2, 2 * pi * (1e9 + 10e6));

% Generate the driving signal
dt = 50e-12;
Ngauss = round(pi/2/system.getRabiFrequency()/dt);
ygauss = exp(-((1:Ngauss) - 0.5*Ngauss).^2/2/(0.15*Ngauss).^2);
ygauss = ygauss * Ngauss/sum(ygauss);
yzeros = zeros(1, Ngauss);
ydetun = ones(1, Ngauss) * (1 - 2e-3) * system.getChargingEnergy();
ypiby2 = ones(1, Ngauss) * pi/2;

% The microwave signals
Nop = 7 * Ngauss;
t = (1:Nop) * dt;
microwave_control_1 = [ygauss, yzeros, ygauss, yzeros, ygauss, yzeros, ygauss] .* cos(system.getLarmorFrequency(1) * t ...
                    + [yzeros, yzeros, ypiby2, yzeros, yzeros, yzeros, ypiby2]);
microwave_control_2 = [yzeros, yzeros, yzeros, 2*ygauss, yzeros, yzeros, yzeros] .* cos(system.getLarmorFrequency(2) * t ...
                    + [yzeros, yzeros, yzeros, yzeros, yzeros, yzeros, yzeros]);
detuning_control    = [yzeros, ydetun, yzeros, yzeros, yzeros, ydetun, yzeros];

% Plot the eigen energies
figure();
spine.plotEigenEnergies(system);

% Simulate the system
nsim = 0;
Nsim = Nop;
spine.simulate(system.getDimension(), @inHamiltonian, @outOperation, @spine.solvers.solver_expm);

% Plot the result
figure();
plot(t, probability);

function [run, H, timestep] = inHamiltonian()
    global nsim;
    global Nsim;
    global system;
    global microwave_control_1 microwave_control_2 detuning_control;
    global dt;

    if (nsim < Nsim)
        
        % Set the signal at this time instance
        system.setMicrowaveControl(1, microwave_control_1(nsim + 1));
        system.setMicrowaveControl(2, microwave_control_2(nsim + 1));
        system.setDetuningControl(detuning_control(nsim + 1));
        
        % Update the Hamiltonian accordingly
        H = system.updateHamiltonian();
        
        % Provide the current timestep, and continue the simulation
        timestep = dt;
        run = true;
        
    else
        run = false;
        H = [];
        timestep = [];
    end
    
end

function outOperation(U)
    global nsim;
    global Nsim;
    global system;
    global dt;
    global probability;

    % Plot every Nplot-th point
    Nplot = 10;
    
    % Store the measurement probability of the singlet state
    probability(nsim + 1) = abs(U(5,1))^2;
    
    % Show the rotation in the Bloch sphere
    if (mod(nsim, Nplot) == 0)
        system.plot(U, nsim * dt, 1);
        pause(0.001);
    end
    
    % Continue the simulation
    nsim = nsim + 1;
    if (nsim == Nsim)
        
        % Last point, print the unitary matrix
        disp(U);
        
    end
    
end

File: cpp/example_2_spin_1_singlet.cpp
/***************************************************************
* Author: J.P.G. van Dijk
***************************************************************/

// Performance/accuracy settings
//#define MKL					(define at project level)
//#define DOUBLE_PRECISION		(define at project level)

// Includes
#define _USE_MATH_DEFINES
#include <math.h>
#include <iostream>
#include "spine/simulate.h"
#include "spine/fidelity.h"
#include "spine/math.h"
#include "spine/solvers/solver_diagonalization.h"
#include "spine/systems/system_2_spin_1_singlet.h"

// Namespaces
using namespace std;
using namespace spine::math;

// Function prototypes
bool inHamiltonian(realnum * H, realnum * dt);
void outOperation(complex * U);

// Private global variables/constants
static spine::systems::system_2_spin_1_singlet spin_system = spine::systems::system_2_spin_1_singlet();
static const realnum dt = (realnum) 50e-12;
static int Nsim, nsim;
static int Ngauss;
static double * ygauss;

int main(void)
{
	// Create the system
	spin_system.setChargingEnergy((realnum)(2 * M_PI * 100e9));
	spin_system.setRabiFrequency((realnum)(2 * M_PI * 1e6));
	spin_system.setTunnelControl((realnum)(2 * M_PI * 10e6 * sqrt(2)));
	spin_system.setLarmorFrequency(0, (realnum)(2 * M_PI * (1e9 - 10e6)));
	spin_system.setLarmorFrequency(1, (realnum)(2 * M_PI * (1e9 + 10e6)));

	// Generate the driving signal
	Ngauss = (int)(M_PI / 2 / spin_system.getRabiFrequency() / dt);
	ygauss = new double[Ngauss];
	double sum = 0;
	for (int n = 0; n < Ngauss; n++) {
		double x = -2.5 + 5.0 * n / (realnum)(Ngauss - 1);
		ygauss[n] = exp(-x * x / 2.0);
		sum += ygauss[n];
	}
	double scale = Ngauss / sum;
	for (int n = 0; n < Ngauss; n++)
		ygauss[n] = ygauss[n] * scale;

	// Simulate the system
	nsim = 0;
	Nsim = 7*Ngauss;
#ifdef PLOT
	spin_system.plotSetup(Nsim, 0, Nsim * dt);
#endif
	cout << "Simulating..." << endl;
	complex * state = allocComplex(spin_system.getDimension());
	spin_system.initialize(state);
	spine::simulate(spin_system.getDimension(), inHamiltonian, outOperation, spine::solvers::solver_diagonalization, state);
	freeComplex(state);
	cout << "finished!" << endl;
#ifdef PLOT
	Plot::run();
#endif

	delete[] ygauss;
}

bool inHamiltonian(realnum * H, realnum * dt)
{
	if (nsim < Nsim)
	{
		// Set the signal at this time instance
		spin_system.setMicrowaveControl(0);
		spin_system.setDetuningControl(0);
		if (nsim < Ngauss) {
			spin_system.setMicrowaveControl(0, ygauss[nsim] * cos(spin_system.getLarmorFrequency(0) * nsim * ::dt));
		}
		else if (nsim < 2 * Ngauss) {
			spin_system.setDetuningControl((1 - 2e-3) * spin_system.getChargingEnergy());
		}
		else if (nsim < 3 * Ngauss) {
			spin_system.setMicrowaveControl(0, ygauss[nsim - 2 * Ngauss] * cos(spin_system.getLarmorFrequency(0) * nsim * ::dt + M_PI / 2.0));
		}
		else if (nsim < 4 * Ngauss) {
			spin_system.setMicrowaveControl(1, 2 * ygauss[nsim - 3 * Ngauss] * cos(spin_system.getLarmorFrequency(1) * nsim * ::dt));
		}
		else if (nsim < 5 * Ngauss) {
			spin_system.setMicrowaveControl(0, ygauss[nsim - 4 * Ngauss] * cos(spin_system.getLarmorFrequency(0) * nsim * ::dt));
		}
		else if (nsim < 6 * Ngauss) {
			spin_system.setDetuningControl((1 - 2e-3) * spin_system.getChargingEnergy());
		}
		else if (nsim < 7 * Ngauss) {
			spin_system.setMicrowaveControl(0, ygauss[nsim - 6 * Ngauss] * cos(spin_system.getLarmorFrequency(0) * nsim * ::dt + M_PI / 2.0));
		}

		// Update the Hamiltonian accordingly
		spin_system.updateHamiltonian(H);

		// Provide the current timestep, and continue the simulation
		*dt = ::dt;
		return true;
	}
	return false;
}

void outOperation(complex * U)
{
	// Plot every Nplot-th point
	const unsigned int Nplot = 10;

	// Print the measurement probability
#ifdef PLOT
	spin_system.plotAdd(U, nsim * dt);
#endif
	if ((nsim % Nplot) == 0)
	{
		realnum P = abs2(U[4]);
		cout << P << endl;
#ifdef PLOT
		spin_system.plot();
#endif
	}

	// Continue the simulation
	nsim++;
}
